aten/src/ATen/native/cuda/Normalization.cu

```diff
@@ -0,0 +1,444 @@
+    v1 += a.v1;
+    v2 += a.v2;
+    return *this;
+  }
+};
+
+template <typename scalar_t, typename accscalar_t>
+struct SumOp {
+  __device__ SumOp(const PackedTensorAccessor<scalar_t, 3, at::RestrictPtrTraits>& t) : tensor(t) {}
+  __device__ __forceinline__ accscalar_t operator()(int batch, int plane, int n) {
+    return static_cast<accscalar_t>(tensor[batch][plane][n]);
+  }
+  const PackedTensorAccessor<scalar_t, 3, at::RestrictPtrTraits>& tensor;
+};
+
+template <typename scalar_t, typename accscalar_t>
+struct VarOp {
+  __device__ VarOp(accscalar_t m, const PackedTensorAccessor<scalar_t, 3, at::RestrictPtrTraits>& t) : mean(m), tensor(t) {}
+  __device__ __forceinline__ accscalar_t operator()(int batch, int plane, int n) {
+    accscalar_t val = tensor[batch][plane][n];
+    return (val - mean) * (val - mean);
+  }
+  const accscalar_t mean;
+  const PackedTensorAccessor<scalar_t, 3, at::RestrictPtrTraits>& tensor;
+};
+
+template <typename scalar_t, typename accscalar_t>
+struct GradOp {
+  __device__ GradOp(accscalar_t m, const PackedTensorAccessor<scalar_t, 3, at::RestrictPtrTraits>& i, const PackedTensorAccessor<scalar_t, 3, at::RestrictPtrTraits>& g)
+    : mean(m), input(i), grad_output(g) {}
+  __device__ __forceinline__ Float2<scalar_t, accscalar_t> operator()(int batch, int plane, int n) {
+    accscalar_t g = grad_output[batch][plane][n];
+    accscalar_t c = static_cast<accscalar_t>(input[batch][plane][n]) - mean;
+    return Float2<scalar_t, accscalar_t>(g, g * c);
+  }
+  const accscalar_t mean;
+  const PackedTensorAccessor<scalar_t, 3, at::RestrictPtrTraits>& input;
+  const PackedTensorAccessor<scalar_t, 3, at::RestrictPtrTraits>& grad_output;
+};
+
+// Sum across all threads within a warp
+template <typename T>
+static __device__ __forceinline__ T warpSum(T val) {
+#if __CUDA_ARCH__ >= 300
+  for (int i = 0; i < getMSB(WARP_SIZE); ++i) {
+    val += WARP_SHFL_XOR(val, 1 << i, WARP_SIZE);
+  }
+#else
+  __shared__ T values[MAX_BLOCK_SIZE];
+  values[threadIdx.x] = val;
+  __threadfence_block();
+  const int base = (threadIdx.x / WARP_SIZE) * WARP_SIZE;
+  for (int i = 1; i < WARP_SIZE; i++) {
+    val += values[base + ((i + threadIdx.x) % WARP_SIZE)];
+  }
+#endif
+  return val;
+}
+
+template <typename scalar_t, typename accscalar_t>
+static __device__ __forceinline__ Float2<scalar_t, accscalar_t> warpSum(Float2<scalar_t, accscalar_t> value) {
+  value.v1 = warpSum(value.v1);
+  value.v2 = warpSum(value.v2);
+  return value;
+}
+
+// Sum across (batch, x/y/z) applying Op() pointwise
+template<typename scalar_t, typename Op, typename PTA>
+__device__ scalar_t reduce(Op op, PTA tensor, int plane) {
+  scalar_t sum = static_cast<scalar_t>(0);
+  for (int batch = 0; batch < tensor.size(0); ++batch) {
+    for (int x = threadIdx.x; x < tensor.size(2); x += blockDim.x) {
+      sum += op(batch, plane, x);
+    }
+  }
+
+  // sum over NumThreads within a warp
+  sum = warpSum(sum);
```

Comment:
I'm a little bit lost here. I thought `warpSum` already sums all threads in a warp. Why do we need reduce again below?

Reply:
This is unchanged logic from THCUNN, it took me a very close look to understand, too.

It's a two step warp sum to reduce from "one result per thread" to "<=32 results" to 1 result.
What the code does is that it has a first warpSum that reduces from whatever number of threads (<= 1024=32*32) to <= 32 intermediate sums. Then the "warp leader" will put that into shared memory. Then there is another round of reduction with warpSum  where the intermediate sums are reduced.
The alternative pattern is a reduction where you reduce down to <=32 first an then have a single round of reduction. (Which is done e.g. for the WeightNorm code.)
From what I've learned from @mcarilli and @jjsjann123, I think that it is a pattern that may offer some performance advantage, but not that much. In fact, my first stab at having a better backward uses the single reduction pattern.

Reply:
This makes much more sense! Thanks for the explanation. Could you please update the function name and leave some comments here? Thanks! :)